metrics/options.go

```diff
@@ -0,0 +1,65 @@
+package metrics
+
+var (
+	// The Prometheus metrics will be made available on this port:
+	defaultPrometheusListenAddress = ":9000"
+	// This is the endpoint where the Prometheus metrics will be made available ("/metrics" is the default with Prometheus):
+	defaultPrometheusPath = "/metrics"
+	// timingObjectives is the default spread of stats we maintain for timings / histograms:
+	defaultTimingObjectives = map[float64]float64{0.0: 0, 0.5: 0.05, 0.75: 0.04, 0.90: 0.03, 0.95: 0.02, 0.98: 0.001, 1: 0}
+)
+
+// Option powers the configuration for metrics implementations:
+type Option func(*Options)
+
+// Options for metrics implementations:
+type Options struct {
+	DefaultTags             Tags
+	PrometheusListenAddress string
+	PrometheusPath          string
+	TimingObjectives        map[float64]float64
+}
+
+// NewOptions prepares a set of options:
+func NewOptions(opt ...Option) Options {
+	opts := Options{
+		DefaultTags:             make(Tags),
+		PrometheusListenAddress: defaultPrometheusListenAddress,
+		PrometheusPath:          defaultPrometheusPath,
+		TimingObjectives:        defaultTimingObjectives,
+	}
+
+	for _, o := range opt {
+		o(&opts)
+	}
+
+	return opts
+}
+
+// PrometheusPath is the path that Prometheus metrics will be made available on:
+func PrometheusPath(value string) Option {
+	return func(o *Options) {
+		o.PrometheusPath = value
+	}
+}
+
+// PrometheusListenAddress is the listen address for the Prometheus client:
+func PrometheusListenAddress(value string) Option {
+	return func(o *Options) {
+		o.PrometheusListenAddress = value
+	}
+}
+
+// DefaultTags will be added to every metric:
+func DefaultTags(value Tags) Option {
+	return func(o *Options) {
+		o.DefaultTags = value
+	}
+}
+
+// TimingObjectives defines the desired spread of statistics for histogram / timing metrics:
+func TimingObjectives(value map[float64]float64) Option {
```

Comment:
I dont understand what TimingObjectives are, is there another name for this?

Reply:
This one is difficult to describe succinctly. It is where we define the list of quantiles and their respective error ranges. Each one of these becomes a statistical metric on timings.

Our defaults are `{0.0: 0, 0.5: 0.05, 0.75: 0.04, 0.90: 0.03, 0.95: 0.02, 0.98: 0.001, 1: 0}`

This means that for each timing histogram metric we'll get 50th, 75th, 90th, 95th, 98th, and 100 quantiles (along with min/max/average/count).

Annoyingly this is quite fiddly with Prometheus, less so with other metrics implementations. We could just leave this out (as a hard-coded default), but I guarantee that it will annoy _someone_ if they can't configure it.

From the Prometheus client repo:
```
	// Objectives defines the quantile rank estimates with their respective
	// absolute error. If Objectives[q] = e, then the value reported for q
	// will be the φ-quantile value for some φ between q-e and q+e.  The
	// default value is an empty map, resulting in a summary without
	// quantiles.
```

Reply:
Can you provide a comment on this function to provide that clarity. More helpful with some explanation